src/Agent.Worker/Worker.cs

```diff
@@ -10,103 +10,76 @@ namespace Microsoft.VisualStudio.Services.Agent.Worker
     [ServiceLocator(Default = typeof(Worker))]
     public interface IWorker : IAgentService
     {
-        Task<int> RunAsync(string pipeIn, string pipeOut, CancellationTokenSource tokenSource);
+        Task RunAsync(string pipeIn, string pipeOut, CancellationTokenSource hostTokenSource);
     }
 
     public sealed class Worker : AgentService, IWorker
     {
-        //allow up to 30sec for the first message to be received
         private readonly TimeSpan WorkerStartTimeout = TimeSpan.FromSeconds(30);
 
-        public async Task<int> RunAsync(string pipeIn, string pipeOut, CancellationTokenSource tokenSource)
+        public async Task RunAsync(string pipeIn, string pipeOut, CancellationTokenSource hostTokenSource)
         {
+            // Validate args.
+            ArgUtil.NotNullOrEmpty(pipeIn, nameof(pipeIn));
+            ArgUtil.NotNullOrEmpty(pipeOut, nameof(pipeOut));
+            ArgUtil.NotNull(hostTokenSource, nameof(hostTokenSource));
+            var jobRunner = HostContext.GetService<IJobRunner>();
+
             using (var channel = HostContext.CreateService<IProcessChannel>())
```

Comment:
I ran into an issue with the unit tests when I refactored the exit code out of job runner. I studied this class for a while and ended up deciding to remove the loop. There was too many different cancellation tokens, threads, and local variables for the state machine to be clearly followed. I removed the while loop and coded it in a straight line. It could be argued that it's less flexible now that it's not a loop. Yet even the loop made assumptions about the flow in which different types of messages were received.